### ISSUE TITLE
Splitting model across 2 GPUs leads to OOM

### ISSUE THREAD

**Issue by yselivonchyk** [cybertronai/gradient-checkpointing#38]
I am running a UNet based model on a single GPU, using gradients_speed.
When splitting the same model across 2 GPUs training runs out with OOM before even starting.

Same model runs fine on 2 GPUs with regular gradients.

What would be a good place to start investigating this issue? What can be causing that?

**Comment by yselivonchyk** [comment]
More info:

After studying the issue further, I found out that some operations migrated to GPU:0 from GPU:1 after using GradientCheckpointing:
`I tensorflow/core/common_runtime/placer.cc:886] training/Adam/gradients_7/training/Adam/batch_normalization_13/moments/mean_grad/Prod_1: (Prod)/job:localhost/replica:0/task:0/device:GPU:0`
where batch_normalization_13 and all it's inputs are placed on GPU:1

**Comment by yaroslavvb** [comment]
Are you sure you need 2 GPUs? Gradient checkpointing makes the memory requirement a lot smaller so it could fit into 1 GPU.

There are 2 places in the code that set device, search for "  op._set_device(origin_op.node_def.device)"

ie this line:

```
   for origin_op, op in info._transformed_ops.items():
        op._set_device(origin_op.node_def.device)
```
The the way to debug would be to figure out if the optimizer ops are explicitly being set to GPU:0 in one of those 2 places.

Otherwise, the problem could be that their device attribute is unset, in which case tensorflow runtime will puts them on GPU:0.

In the latter case, the question is -- who creates those extra optimizer ops?

If they are created by graph_editor, then they should be included in the "set_device" loop

**Comment by yselivonchyk** [comment]
There are 2 sets of ops created without explicit device:
- tf.stop_gradients() calls, which was very straight forward to fix.
- ops constructed by call to tf_gradients(boundary, ...). Operations produced by this call seems to be assigned to a different `gradients_%d` namespace

**Comment by yselivonchyk** [comment]
The issue originates from tf.gradients(). It doesn't assign the corresponding device to all the ops that it creates, such as `gradients/conv*/*/Conv2DBackpropInpum, gradients/*/Conv2DBackpropFilter`. 

As a workaround I used explicit device assignment for tf.stopgradients and tf_gradients calls within the script. Yet, it does not fully adresses the issue.

Close reason: the issue does not originate from the package

**Comment by yaroslavvb** [comment]
Do you having adding `with tf.device()` statements around `tf_gradients(boundary` calls would've fixed the issue?

**Comment by yselivonchyk** [comment]
I am setting` tf.device()` for both calls of `tf_gradients` to corresponding `boundary[0].op.node_def.device`. If before gradients were spilling over for every op in the graph, now they will spill over only for supgraphs that span across multiple GPUs. So, it mitigates the issue to a large extend, but does not fix exactly. I might forsee this getting out of hand again if subgraphs are too are large or having multiple boundaries (which is unlikely to happen).

**Comment by yaroslavvb** [comment]
Thanks for getting to the root of this. This fix would be useful for other users, let me know if you'd like to submit a PR, or otherwise if you send me a copy of your `memory_saving_gradients.py` I could incorporate this change.

**Comment by yselivonchyk** [comment]
i'll do the PR

**Comment by yselivonchyk** [comment]
https://github.com/cybertronai/gradient-checkpointing/pull/43
